pygmt/io.py

```diff
@@ -0,0 +1,36 @@
+"""
+PyGMT i/o utilities.
+"""
+import xarray as xr
+
+
+def process_output_grid(grid_name, tmpfile_name):
```

Comment:
Do we want this to be a public facing function? If this is meant for internal use only, I suggest renaming the function to `_process_output_grid` as per https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables.

However, if we do want this to be public-facing (i.e. users can do `pygmt.io.process_output_grid`), I would suggest another name like `open_datagrid` (or similar), and a completely different API implementation that follows in the footsteps of [`rioxarray.open_rasterio`](https://corteva.github.io/rioxarray/stable/rioxarray.html#rioxarray.open_rasterio).

There is also the option of creating a custom pygmt 'engine' in `xarray`, see e.g. https://github.com/corteva/rioxarray/pull/281, but I'm getting ahead of myself here. I'm just thinking more long-term xarray integration plans (xref #499).

Reply:
While I think a open_grid() function or pygmt engine in xarray would be excellent longterm (analogous to GMT.jl's [gdalread_function](https://www.generic-mapping-tools.org/GMT.jl/dev/#GMT.gdalread)), this function is not for that purpose. So I will rename appropriately.

Reply:
Ideally, we should avoid writing the grid to a temporary file. Instead, we should define a ["GMT_GRID" class](https://docs.generic-mapping-tools.org/latest/api.html#gmt-grids), output the grid into a GMT_GRID object, but we may also need a function to convert a GMT_GRID object to xarray.DataArray.

I tried it before but didn't succeed due to my limited knowledge about ctypes.

Reply:
Yes, I can see this being a better option. I do not have any time to look into a more robust refactoring now, so unless someone else has free time soon I suggest we still merge this PR and open a new issue for your idea.

Reply:
Agree with merging this PR (after renaming the function to have a leading underscore). The `GMT_GRID` class can wait, from experience it's probably not a trivial thing to add (xref https://github.com/GenericMappingTools/pygmt/pull/520#discussion_r454096421) though probably something we need to think about longer term.

Reply:
Some clarification would be helpful- The PEP8 link is for methods and instances but this is a function. The section for functions does not have similar guidance and most analogous functions in utils.py do not use a leading underscore. Should I still add one here?

Reply:
Sorry, a better link is probably https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces. To quote (emphasis mine):

> ..., internal interfaces (packages, modules, classes, **functions**, attributes or other names) should still be **prefixed with a single leading underscore**.

So you can just rename it like so:

```suggestion
def _process_output_grid(grid_name, tmpfile_name):
```
and make sure to apply it to the other lines in this PR.

Reply:
As the variable names `grid_name` and `tempfile_name` make little sense in this function, I'm wondering if we can rewrite the function to:
```
def process_output_grid(bool_flag)
```
and call it this way `process_output_grid(outgrid == tmpfile.name)` in other functions.

Reply:
Do you mean something like:

```python
def process_output_grid(outgrid, bool_flag):
```

and used like `return process_output_grid(outgrid, bool_flag=outgrid == tmpfile.name)`?

To be honest, I did think of using a one-liner like `return load_datagrid(outgrid) if outgrid == tmpfile.name else None`, where the `load_datagrid` function would just have the `with xr.open_dataarray()` code-block. The `load_datagrid` function could then be public-facing, and is essentially similar to [`xarray.load_dataarray`](http://xarray.pydata.org/en/v0.19.0/generated/xarray.load_dataarray.html) but with an extra `_ = result.gmt  # load GMTDataArray accessor information` line.

Reply:
I like the `return load_datagrid(outgrid) if outgrid == tmpfile.name else None` code. Actually I feel it's a little weird to check `if outgrid == tmpfile.name` in the `process_output_grid` function, which is not easy to understand for new wrapper contributors.

Reply:
Let's see what Meghan thinks when she comes back tomorrow. If we do decide to go with `load_datagrid` though, the code here can be replaced too :grin::

https://github.com/GenericMappingTools/pygmt/blob/ebbb0cc0045454f05db5abdfb3bcb5e1f18e96a7/pygmt/datasets/earth_relief.py#L136-L138

Reply:
Sounds good, I'll try out your implementation idea. Going to convert to draft for now and come back to this after more workshop prep work.